core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -19,14 +19,45 @@
 
 package org.elasticsearch.common.path;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.unmodifiableMap;
 
 public class PathTrie<T> {
 
+    enum TrieMatchingMode {
```

Comment:
I wonder, if we would always retrieve all matching one (recall optimized) and then sort once we have them by the conditions we have here would that be simpler code wise? Also down the road, not in this PR should we use an FST <String, int> that we could score by some predifined value to get the sorted order by default and safe a boat load of objects? the output of the fst would then just point into an array to get the handler...